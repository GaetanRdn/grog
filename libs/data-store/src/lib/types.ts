import { Observable } from 'rxjs';

export type Concrete<Type> = {
  [Property in keyof Type]-?: Type[Property];
};

export type OptionalPropertiesOf<Type> = {
  // eslint-disable-next-line @typescript-eslint/ban-types
  [K in keyof Type]-?: {} extends Pick<Type, K> ? K : never;
}[keyof Type];

export type MandatoryPropertiesOf<Type> = Omit<
  Type,
  OptionalPropertiesOf<Type>
>;

export interface StorableEntity<EntityType> {
  entity: EntityType;
  change: Partial<EntityType> | null;
  readonly state: StoredEntityState;
  readonly stateChanged$: Observable<StoredEntityState>;
  delete: () => void;
}

export type StoredEntityState =
  | 'initialized'
  | 'deleted'
  | 'updated'
  | 'partial'
  | 'created';

export interface StoreConfig<EntityType> {
  extractIdentifier: keyof EntityType | (keyof EntityType)[];
}
