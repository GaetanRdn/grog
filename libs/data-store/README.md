# data-store

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test data-store` to execute the unit tests.
