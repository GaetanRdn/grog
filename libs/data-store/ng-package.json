{
  "$schema": "../../node_modules/ng-packagr/ng-package.schema.json",
  "dest": "../../dist/libs/data-store",
  "lib": {
    "entryFile": "src/index.ts"
  }
}
