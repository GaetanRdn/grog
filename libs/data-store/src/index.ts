export * from './lib/data-store.module';
