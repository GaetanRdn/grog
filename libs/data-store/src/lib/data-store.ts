import { Injectable } from '@angular/core';
import { BehaviorSubject, Observable } from 'rxjs';
import { StoredEntity } from './stored-entity';
import {
  Concrete,
  MandatoryPropertiesOf,
  StorableEntity,
  StoreConfig,
  StoredEntityState,
} from './types';
import { deepCopy } from './utils';

@Injectable({
  providedIn: 'root',
})
export class DataStore<
  EntityType,
  Identifier extends keyof Concrete<EntityType> = keyof MandatoryPropertiesOf<EntityType>
> {
  private _store: BehaviorSubject<StoredEntity<EntityType>[]> =
    new BehaviorSubject<StoredEntity<EntityType>[]>([]);

  private _identifierKeys: (keyof EntityType)[];

  get changed$(): Observable<StorableEntity<EntityType>[]> {
    return this._store.asObservable();
  }

  constructor(private _config: StoreConfig<EntityType>) {
    this._identifierKeys = Array.isArray(this._config.extractIdentifier)
      ? this._config.extractIdentifier
      : [this._config.extractIdentifier];
  }

  public store(...entities: EntityType[]): DataStore<EntityType, Identifier> {
    this._store.next([
      ...this._store.getValue(),
      ...entities.map(
        (entity: EntityType) => new StoredEntity(deepCopy(entity), this)
      ),
    ]);

    return this;
  }

  public get(
    identifier: Pick<EntityType, Identifier>
  ): StorableEntity<EntityType> | null {
    return (
      this._store
        .getValue()
        .find((storedEntity: StoredEntity<EntityType>) =>
          this.sameEntity(identifier, storedEntity.entity)
        ) ?? null
    );
  }

  public delete(identifier: Pick<EntityType, Identifier>): void {
    this.updateState(identifier, 'deleted');
  }

  public update(
    entity: Pick<EntityType, Identifier>,
    change: Partial<EntityType> & MandatoryPropertiesOf<EntityType>
  ): void {
    this.updateState(entity, 'updated', change);
  }

  public partialUpdate(
    entity: Pick<EntityType, Identifier>,
    change: Partial<EntityType>
  ): void {
    this.updateState(entity, 'partial', change);
  }

  public create(entity: MandatoryPropertiesOf<EntityType>): void {
    this._store.next([
      ...this._store.getValue(),
      new StoredEntity(deepCopy(entity) as EntityType, this, 'created'),
    ]);
  }

  private updateState(
    identifier: Pick<EntityType, Identifier>,
    state: StoredEntityState,
    change?: Partial<EntityType>
  ): void {
    this._store.next(
      this._store.getValue().map((storedEntity: StoredEntity<EntityType>) => {
        if (this.sameEntity(identifier, storedEntity.entity)) {
          storedEntity['_state'] = state;

          if (change) {
            storedEntity.change = change;
          }
        }
        return storedEntity;
      })
    );
  }

  private sameEntity(
    identifier: Pick<EntityType, Identifier>,
    entity: EntityType
  ): boolean {
    return this._identifierKeys.every(
      (key: keyof EntityType) => (identifier as EntityType)[key] === entity[key]
    );
  }
}
