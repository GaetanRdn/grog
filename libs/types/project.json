{
  "projectType": "library",
  "root": "libs/types",
  "sourceRoot": "libs/types/src",
  "prefix": "grorg",
  "targets": {
    "build": {
      "executor": "@nrwl/angular:ng-packagr-lite",
      "outputs": ["dist/libs/types"],
      "options": {
        "project": "libs/types/ng-package.json"
      },
      "configurations": {
        "production": {
          "tsConfig": "libs/types/tsconfig.lib.prod.json"
        },
        "development": {
          "tsConfig": "libs/types/tsconfig.lib.json"
        }
      },
      "defaultConfiguration": "production"
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "options": {
        "lintFilePatterns": [
          "libs/types/src/**/*.ts",
          "libs/types/src/**/*.html"
        ]
      }
    }
  },
  "tags": ["scope:shared"]
}
