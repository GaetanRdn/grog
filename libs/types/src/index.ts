export * from './lib/utility-types';
