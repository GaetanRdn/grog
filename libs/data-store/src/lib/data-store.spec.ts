import { DataStore } from './data-store';
import { Person } from './data-store.mocks';
import { StorableEntity } from './types';

describe('DataStore', () => {
  let personStore: DataStore<Person, 'id'>;

  beforeEach(() => {
    personStore = new DataStore({
      extractIdentifier: 'id',
    });
  });

  test('when store 1 entity then check changed$', (doneFn: jest.DoneCallback) => {
    // WHEN
    personStore.store({
      firstName: 'Gaetan',
      lastName: 'REDIN',
    });

    // THEN
    personStore.changed$.subscribe((changed: StorableEntity<Person>[]) => {
      expect(changed).toMatchSnapshot();
      changed.forEach(({ state }: StorableEntity<Person>) =>
        expect(state).toEqual('initialized')
      );
      doneFn();
    });
  });

  test('when store 2 entities at the same time then check changed$', (doneFn: jest.DoneCallback) => {
    // WHEN
    personStore.store(
      {
        firstName: 'Elisa',
        lastName: 'REDIN',
        age: 25,
      },
      {
        firstName: 'Gaetan',
        lastName: 'REDIN',
      }
    );

    // THEN
    personStore.changed$.subscribe((changed: StorableEntity<Person>[]) => {
      expect(changed).toMatchSnapshot();
      changed.forEach(({ state }: StorableEntity<Person>) =>
        expect(state).toEqual('initialized')
      );
      doneFn();
    });
  });

  test('when store 2 entities one after another then check last changed$', (doneFn: jest.DoneCallback) => {
    // WHEN
    personStore
      .store({
        firstName: 'Elisa',
        lastName: 'REDIN',
        age: 25,
      })
      .store({
        firstName: 'Gaetan',
        lastName: 'REDIN',
      });

    // THEN
    personStore.changed$.subscribe((changed: StorableEntity<Person>[]) => {
      expect(changed).toMatchSnapshot();
      changed.forEach(({ state }: StorableEntity<Person>) =>
        expect(state).toEqual('initialized')
      );
      doneFn();
    });
  });

  test('when get entity and empty store then return null', () => {
    expect(personStore.get({ id: 1 })).toBeNull();
  });

  test('when get entity and store contains entity then return it', () => {
    // GIVEN
    personStore.store({
      id: 1,
      firstName: 'Gaetan',
      lastName: 'REDIN',
    });

    // WHEN
    const current: StorableEntity<Person> | null = personStore.get({ id: 1 });

    // THEN
    expect(current).toMatchSnapshot({
      entity: {
        id: 1,
        firstName: 'Gaetan',
        lastName: 'REDIN',
      },
    });
    expect(current?.state).toEqual('initialized');
  });

  test('when delete entity then check changed$', (doneFn: jest.DoneCallback) => {
    // GIVEN
    personStore.store({
      id: 1,
      firstName: 'Gaetan',
      lastName: 'REDIN',
    });

    // WHEN
    personStore.delete({ id: 1 });

    // THEN
    personStore.changed$.subscribe((changed: StorableEntity<Person>[]) => {
      expect(changed).toMatchSnapshot();
      expect(changed[0].state).toEqual('deleted');
      doneFn();
    });
  });

  test('when update entity then check changed$', (doneFn: jest.DoneCallback) => {
    // GIVEN
    const entity: Person = {
      id: 1,
      firstName: 'Gaetan',
      lastName: 'REDIN',
    };
    personStore.store(entity);

    // WHEN
    entity.age = 31;
    personStore.update({ id: 1 }, entity);

    // THEN
    personStore.changed$.subscribe((changed) => {
      expect(changed).toMatchSnapshot();
      expect(changed[0].state).toEqual('updated');
      doneFn();
    });
  });

  test('when create entity then check changed$', (doneFn: jest.DoneCallback) => {
    // GIVEN
    const entity: Person = {
      firstName: 'Gaetan',
      lastName: 'REDIN',
    };

    // WHEN
    personStore.create(entity);

    // THEN
    personStore.changed$.subscribe((changed) => {
      expect(changed).toMatchSnapshot();
      expect(changed[0].state).toEqual('created');
      doneFn();
    });
  });

  test('when partialUpdate entity then check changed$', (doneFn: jest.DoneCallback) => {
    // GIVEN
    const entity: Person = {
      id: 1,
      firstName: 'Gaetan',
      lastName: 'REDIN',
    };
    personStore.store(entity);

    // WHEN
    personStore.partialUpdate({ id: 1 }, { age: 31 });

    // THEN
    personStore.changed$.subscribe((changed: StorableEntity<Person>[]) => {
      expect(changed).toMatchSnapshot();
      expect(changed[0].state).toEqual('partial');
      doneFn();
    });
  });

  test.todo('when flush store then check changed$');
});
