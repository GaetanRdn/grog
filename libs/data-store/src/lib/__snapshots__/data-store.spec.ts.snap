// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataStore when create entity then check changed$ 1`] = `
Array [
  StoredEntity {
    "change": null,
    "entity": Object {
      "firstName": "Gaetan",
      "lastName": "REDIN",
    },
  },
]
`;

exports[`DataStore when delete entity then check changed$ 1`] = `
Array [
  StoredEntity {
    "change": null,
    "entity": Object {
      "firstName": "Gaetan",
      "id": 1,
      "lastName": "REDIN",
    },
  },
]
`;

exports[`DataStore when get entity and store contains entity then return it 1`] = `
Object {
  "change": null,
  "entity": Object {
    "firstName": "Gaetan",
    "id": 1,
    "lastName": "REDIN",
  },
}
`;

exports[`DataStore when partialUpdate entity then check changed$ 1`] = `
Array [
  StoredEntity {
    "change": Object {
      "age": 31,
    },
    "entity": Object {
      "firstName": "Gaetan",
      "id": 1,
      "lastName": "REDIN",
    },
  },
]
`;

exports[`DataStore when store 1 entity then check changed$ 1`] = `
Array [
  StoredEntity {
    "change": null,
    "entity": Object {
      "firstName": "Gaetan",
      "lastName": "REDIN",
    },
  },
]
`;

exports[`DataStore when store 2 entities at the same time then check changed$ 1`] = `
Array [
  StoredEntity {
    "change": null,
    "entity": Object {
      "age": 25,
      "firstName": "Elisa",
      "lastName": "REDIN",
    },
  },
  StoredEntity {
    "change": null,
    "entity": Object {
      "firstName": "Gaetan",
      "lastName": "REDIN",
    },
  },
]
`;

exports[`DataStore when store 2 entities one after another then check last changed$ 1`] = `
Array [
  StoredEntity {
    "change": null,
    "entity": Object {
      "age": 25,
      "firstName": "Elisa",
      "lastName": "REDIN",
    },
  },
  StoredEntity {
    "change": null,
    "entity": Object {
      "firstName": "Gaetan",
      "lastName": "REDIN",
    },
  },
]
`;

exports[`DataStore when update entity then check changed$ 1`] = `
Array [
  StoredEntity {
    "change": Object {
      "age": 31,
      "firstName": "Gaetan",
      "id": 1,
      "lastName": "REDIN",
    },
    "entity": Object {
      "firstName": "Gaetan",
      "id": 1,
      "lastName": "REDIN",
    },
  },
]
`;
